File: .tm-cache/process.js
const fs = require('fs');
const m = JSON.parse(fs.readFileSync('.tm-cache/cg-markets.json','utf8'));
const t = JSON.parse(fs.readFileSync('.tm-cache/cg-trending.json','utf8'));

console.log("markets count:", m.length);
console.log("trending count:", (t.coins||[]).length);

const STABLE = new Set(['tether','usd-coin','dai','first-digital-usd','usde','ethena-usde','tusd','true-usd',
  'usdd','pyusd','paypal-usd','fdusd','paxos-standard','paxg','pax-gold','usds','sky-dollar','binance-usd',
  'frax','usdt0','blackrock-usd','ethena-staked-usde','usdtb','susds','susde','ondo-us-dollar-yield',
  'global-dollar','falcon-finance','resolv-usr','binance-bridged-usdt-bnb-smart-chain','usdb','meta-usd']);
const WRAPPED = new Set(['wrapped-bitcoin','weth','wrapped-steth','staked-ether','wrapped-beacon-eth',
  'coinbase-wrapped-btc','wrapped-eeth','rocket-pool-eth','mantle-staked-ether','renzo-restaked-eth',
  'kelp-dao-restaked-eth','lombard-staked-btc','solv-btc','jupiter-staked-sol','binance-staked-sol','msol',
  'jito-staked-sol','liquid-staked-ethereum','bridged-wrapped-steth-scroll','wrapped-eth','wbeth',
  'ether-fi-staked-eth','wrapped-avax','tbtc','clbtc']);

function isStable(c){
  const sym=(c.symbol||'').toUpperCase(), name=(c.name||'').toLowerCase(), id=c.id||'';
  if(STABLE.has(id)) return true;
  if(name.includes('stablecoin')) return true;
  if(sym.startsWith('USD')||sym.startsWith('EUR')||sym.startsWith('GBP')) return true;
  const p=c.current_price||0, ch=c.price_change_percentage_24h;
  if(p>=0.95&&p<=1.05&&ch!=null&&Math.abs(ch)<0.5) return true;
  return false;
}

const filtered = m.filter(c=>{
  if(isStable(c)) return false;
  if(WRAPPED.has(c.id)) return false;
  if((c.total_volume||0)<1_000_000) return false;
  if(c.price_change_percentage_24h==null) return false;
  return true;
});
console.log("filtered survivors:", filtered.length);

const top100 = filtered.slice(0,100);
const green = top100.filter(c=>(c.price_change_percentage_24h||0)>0).length;
const top50 = filtered.slice(0,50).map(c=>c.price_change_percentage_24h||0).sort((a,b)=>a-b);
const med50 = top50.length%2 ? top50[(top50.length-1)/2] : (top50[top50.length/2-1]+top50[top50.length/2])/2;
const find=id=>m.find(c=>c.id===id);
const fmtMajor=c=>c?`${c.symbol.toUpperCase()} $${(c.current_price).toLocaleString('en-US',{maximumFractionDigits:0})} ${(c.price_change_percentage_24h||0).toFixed(1)}% / 7d ${(c.price_change_percentage_7d_in_currency||0).toFixed(1)}%`:'n/a';
console.log(`PULSE: ${green}/${top100.length} top-100 green, median top-50 24h ${med50>=0?'+':''}${med50.toFixed(2)}%`);
console.log("BTC:", fmtMajor(find('bitcoin')));
console.log("ETH:", fmtMajor(find('ethereum')));
console.log("SOL:", fmtMajor(find('solana')));

const by24=[...filtered].sort((a,b)=>(a.price_change_percentage_24h||0)-(b.price_change_percentage_24h||0));
const losers=by24.slice(0,10);
const winners=by24.slice(-10).reverse();

function capAbbr(v){if(v==null)return'n/a';if(v>=1e9)return'$'+(v/1e9).toFixed(1)+'B';if(v>=1e6)return'$'+(v/1e6).toFixed(0)+'M';return'$'+(v/1e3).toFixed(0)+'K';}
function priceFmt(p){if(p==null)return'n/a';if(p>=1)return'$'+Number(p.toPrecision(5)).toLocaleString('en-US');if(p>=0.01)return'$'+p.toFixed(4);return'$'+p.toFixed(6);}

const trendingSyms=new Set();
const trending=(t.coins||[]).slice(0,7).map(it=>{
  const d=it.item, sym=(d.symbol||'').toUpperCase();
  trendingSyms.add(sym);
  const data=d.data||{};
  const ch=(data.price_change_percentage_24h||{}).usd;
  return {name:d.name,sym,rank:d.market_cap_rank,price:data.price,ch24:ch};
});

function tags(c,isWinner){
  const out=[],sym=c.symbol.toUpperCase();
  const ch24=c.price_change_percentage_24h||0, ch7=c.price_change_percentage_7d_in_currency||0;
  const rank=c.market_cap_rank||9999, mcap=c.market_cap||0, vol=c.total_volume||0;
  const volmcap=mcap?vol/mcap:0, inTrend=trendingSyms.has(sym);
  if(inTrend&&ch24>0&&isWinner===true) out.push('TRENDING+UP');
  if(inTrend&&ch24<0&&isWinner===false) out.push('TRENDING+DOWN');
  if(ch24>15&&ch7>25) out.push('BREAKOUT');
  if(ch24>20&&ch7<0) out.push('FADE');
  if(ch24<-10&&volmcap>0.25) out.push('CAPITULATION');
  if(rank>150&&ch24>30) out.push('PUMP-RISK');
  if(mcap&&mcap<50_000_000) out.push('MICROCAP');
  if(rank<=20) out.push('MAJOR');
  return out.slice(0,2);
}
function show(c,isWinner){
  const sym=c.symbol.toUpperCase();
  const ch24=c.price_change_percentage_24h||0,ch7=c.price_change_percentage_7d_in_currency||0,ch1=c.price_change_percentage_1h_in_currency||0;
  return `${sym} (${c.name}) ${priceFmt(c.current_price)} ${ch24>=0?'+':''}${ch24.toFixed(1)}% / 7d ${ch7>=0?'+':''}${ch7.toFixed(1)}% / 1h ${ch1>=0?'+':''}${ch1.toFixed(1)}% • ${capAbbr(c.total_volume)} / #${c.market_cap_rank} ${JSON.stringify(tags(c,isWinner))}`;
}
console.log("\n=== WINNERS ===");
winners.forEach(c=>console.log(show(c,true)));
console.log("\n=== LOSERS ===");
losers.forEach(c=>console.log(show(c,false)));
console.log("\n=== TRENDING ===");
trending.forEach(tr=>console.log(`${tr.name} (${tr.sym}) #${tr.rank} ${priceFmt(tr.price)} 24h ${tr.ch24!=null?(tr.ch24>=0?'+':'')+tr.ch24.toFixed(1)+'%':'n/a'}`));

File: .tm-cache/process.py
import json

m = json.load(open('.tm-cache/cg-markets.json'))
t = json.load(open('.tm-cache/cg-trending.json'))

print("markets count:", len(m))
print("trending count:", len(t.get('coins', [])))

# --- Filter ---
STABLE_IDS = {'tether','usd-coin','dai','first-digital-usd','usde','ethena-usde','tusd','true-usd',
              'usdd','pyusd','paypal-usd','fdusd','paxos-standard','paxg','pax-gold','usds','sky-dollar',
              'binance-usd','frax','usdt0','blackrock-usd','ethena-staked-usde','usdtb','susds','susde',
              'ondo-us-dollar-yield','global-dollar','falcon-finance','resolv-usr','wrapped-eeth'}
WRAPPED_IDS = {'wrapped-bitcoin','weth','wrapped-steth','staked-ether','wrapped-beacon-eth',
               'coinbase-wrapped-btc','binance-bridged-usdt-bnb-smart-chain','wrapped-eeth',
               'rocket-pool-eth','mantle-staked-ether','renzo-restaked-eth','kelp-dao-restaked-eth',
               'lombard-staked-btc','solv-btc','jupiter-staked-sol','binance-staked-sol','msol',
               'jito-staked-sol','liquid-staked-ethereum','bridged-wrapped-steth-scroll'}

def is_stable(c):
    sym = (c.get('symbol') or '').upper()
    name = (c.get('name') or '').lower()
    cid = c.get('id') or ''
    if cid in STABLE_IDS: return True
    if 'stablecoin' in name or 'usd' in name and 'yield' in name: return True
    if sym.startswith('USD') or sym.startswith('EUR') or sym.startswith('GBP'): return True
    # near-peg detection: price ~1.0 with tiny moves
    p = c.get('current_price') or 0
    ch = c.get('price_change_percentage_24h')
    if 0.95 <= p <= 1.05 and ch is not None and abs(ch) < 0.5:
        return True
    return False

filtered = []
for c in m:
    if is_stable(c): continue
    if c.get('id') in WRAPPED_IDS: continue
    vol = c.get('total_volume') or 0
    if vol < 1_000_000: continue
    if c.get('price_change_percentage_24h') is None: continue
    filtered.append(c)

print("filtered survivors:", len(filtered))

# --- Market pulse ---
top100 = filtered[:100] if len(filtered) >= 100 else filtered
green = sum(1 for c in top100 if (c.get('price_change_percentage_24h') or 0) > 0)
top50 = filtered[:50]
import statistics
med50 = statistics.median([c.get('price_change_percentage_24h') or 0 for c in top50])
btc = next((c for c in m if c['id']=='bitcoin'), None)
eth = next((c for c in m if c['id']=='ethereum'), None)
sol = next((c for c in m if c['id']=='solana'), None)
def fmt_major(c):
    return f"{c['symbol'].upper()} ${c['current_price']:,.0f} {c.get('price_change_percentage_24h',0):+.1f}% / 7d {c.get('price_change_percentage_7d_in_currency') or 0:+.1f}%"
print(f"PULSE: {green}/{len(top100)} top-100 green, median top-50 24h {med50:+.2f}%")
print("BTC:", fmt_major(btc) if btc else "n/a")
print("ETH:", fmt_major(eth) if eth else "n/a")
print("SOL:", fmt_major(sol) if sol else "n/a")

# --- Sort winners/losers ---
by24 = sorted(filtered, key=lambda c: c.get('price_change_percentage_24h') or 0)
losers = by24[:10]
winners = list(reversed(by24[-10:]))

def cap_abbr(v):
    if v is None: return "n/a"
    if v >= 1e9: return f"${v/1e9:.1f}B"
    if v >= 1e6: return f"${v/1e6:.0f}M"
    return f"${v/1e3:.0f}K"

def price_fmt(p):
    if p is None: return "n/a"
    if p >= 1: return f"${p:,.4g}"
    if p >= 0.01: return f"${p:.4f}"
    return f"${p:.6f}"

# trending set
trending_syms = set()
trending = []
for item in t.get('coins', [])[:7]:
    d = item['item']
    sym = (d.get('symbol') or '').upper()
    trending_syms.add(sym)
    data = d.get('data') or {}
    ch = (data.get('price_change_percentage_24h') or {}).get('usd')
    trending.append({'name': d.get('name'), 'sym': sym, 'rank': d.get('market_cap_rank'),
                     'price': data.get('price'), 'ch24': ch})

def tags(c, is_winner=None):
    out = []
    sym = c['symbol'].upper()
    ch24 = c.get('price_change_percentage_24h') or 0
    ch7 = c.get('price_change_percentage_7d_in_currency') or 0
    rank = c.get('market_cap_rank') or 9999
    mcap = c.get('market_cap') or 0
    vol = c.get('total_volume') or 0
    volmcap = (vol/mcap) if mcap else 0
    in_trend = sym in trending_syms
    if in_trend and ch24 > 0 and is_winner: out.append('TRENDING+UP')
    if in_trend and ch24 < 0 and is_winner is False: out.append('TRENDING+DOWN')
    if ch24 > 15 and ch7 > 25: out.append('BREAKOUT')
    if ch24 > 20 and ch7 < 0: out.append('FADE')
    if ch24 < -10 and volmcap > 0.25: out.append('CAPITULATION')
    if rank > 150 and ch24 > 30: out.append('PUMP-RISK')
    if mcap and mcap < 50_000_000: out.append('MICROCAP')
    if rank <= 20: out.append('MAJOR')
    return out[:2]

def line(c, is_winner):
    sym = c['symbol'].upper()
    ch24 = c.get('price_change_percentage_24h') or 0
    ch7 = c.get('price_change_percentage_7d_in_currency') or 0
    ch1 = c.get('price_change_percentage_1h_in_currency') or 0
    return {
        'sym': sym, 'name': c['name'], 'rank': c.get('market_cap_rank'),
        'price': price_fmt(c.get('current_price')),
        'ch24': ch24, 'ch7': ch7, 'ch1': ch1,
        'vol': cap_abbr(c.get('total_volume')), 'mcap': cap_abbr(c.get('market_cap')),
        'tags': tags(c, is_winner)
    }

print("\n=== WINNERS ===")
for c in winners:
    l = line(c, True)
    print(f"{l['sym']} ({l['name']}) {l['price']} {l['ch24']:+.1f}% / 7d {l['ch7']:+.1f}% / 1h {l['ch1']:+.1f}% • {l['vol']} / #{l['rank']} {l['tags']}")

print("\n=== LOSERS ===")
for c in losers:
    l = line(c, False)
    print(f"{l['sym']} ({l['name']}) {l['price']} {l['ch24']:+.1f}% / 7d {l['ch7']:+.1f}% / 1h {l['ch1']:+.1f}% • {l['vol']} / #{l['rank']} {l['tags']}")

print("\n=== TRENDING ===")
for tr in trending:
    pr = price_fmt(tr['price'])
    ch = f"{tr['ch24']:+.1f}%" if tr['ch24'] is not None else "n/a"
    extra = ''
    if tr['sym'] in trending_syms: pass
    print(f"{tr['name']} ({tr['sym']}) #{tr['rank']} {pr} 24h {ch}")
